ci: doc: pin Python dependencies to specific versions/hashes

kartben opened the pull request:
Use secure install for Python documentation dependencies.

kartben committed:
ci: doc: pin Python dependencies to specific versions/hashes

Use secure install for ̦Python documentation dependencies.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben committed:
ci: doc: update documentation dependencies to include coverxygen

Removed coverxygen installation from the CI workflow and added it to the
requirements.txt/.in files and make it easier for anyone to potentially
use it locally.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben committed:
ci: github: update dependabot.yml to manage documentation dependencies

Added a new configuration for pip dependencies in the doc/
Groups all updates into one PR.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

pdgendt reviewed (dismissed):
Yes please!

pdgendt commented on .github/dependabot.yml:
```diff
@@ -11,3 +11,15 @@ updates:
         patterns:
           - "*"
+
+  - package-ecosystem: "pip"
```
Is it `pip` or `pip-compile`?

kartben replied:
it's `pip` -- https://docs.github.com/en/code-security/dependabot/ecosystems-supported-by-dependabot/supported-ecosystems-and-repositories#supported-ecosystems-and-repositories

<img width="853" alt="image" src="https://github.com/user-attachments/assets/35e41b7c-e1b4-4358-b650-b09d0e27b17c" />

pdgendt commented:
The job `doc-build-pdf` also has a `install-pip` step:

```yml
    - name: install-pip
      run: |
        pip install -r doc/requirements.txt
        pip install west==${WEST_VERSION}
        pip install cmake==${CMAKE_VERSION}
```

Which has a different construct for the package versions, what's the rationale here, and why can't we simply add these in `doc/requirements.in`?

nashif commented on doc/requirements.txt:
```diff
@@ -1,26 +1,722 @@
-# DOC: used to generate docs
+#
+# This file is autogenerated by pip-compile with Python 3.13
```
what version are we using in CI? Should this match the same version used in CI?

pdgendt replied:
I'm not fluent in ruby, but AFAICS dependabot parses this line to match the python version.

Version parsing:
https://github.com/dependabot/dependabot-core/blob/0405a99f095c792385a61be4f0f11911b81a86e8/python/lib/dependabot/python/file_parser/python_requirement_parser.rb#L92
Matching:
https://github.com/dependabot/dependabot-core/blob/0405a99f095c792385a61be4f0f11911b81a86e8/python/lib/dependabot/python/language_version_manager.rb#L90

This kind of information lacks in the documentation it seems.

nashif replied:
the question is unrelated to dependabot , wondering if this would be a problem to generate hashes using one version and to be using a different version when installing the packages, i.e. there might be incompatibilities.

kartben replied:
I think we're good. The version is there to help dependabot generate requirements.txt files that are "consistent" (i.e. no unnecessary diffs). The actual hashes have nothing to do with it AFAICT and any pip 8+ version will be happy with the format of the file (https://pip.pypa.io/en/stable/topics/secure-installs/#hash-checking-mode)

kartben replied:
re-gen'd the requirements.txt with python 3.12 (Python version used in our docker image)

kartben commented:
> The job `doc-build-pdf` also has a `install-pip` step:
> 
> ```yml
>     - name: install-pip
>       run: |
>         pip install -r doc/requirements.txt
>         pip install west==${WEST_VERSION}
>         pip install cmake==${CMAKE_VERSION}
> ```
> 
> Which has a different construct for the package versions, what's the rationale here, 

Yeah I need to look into this as it seems to actually be failing anyway 
https://github.com/kartben/zephyr/actions/runs/13961212510/job/39082783857

> and why can't we simply add these in `doc/requirements.in`?

Yeah this needs some cleanup. Adding cmake to the requirements file would likely be problematic for folks install dependencies locally and already having CMake? I will install it as a Linux package

kartben committed:
ci: doc: github: check Doxygen MD5 checksum

Add a check that we're installing a legit Doxygen binary

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben committed:
ci: doc: github: Build PDF from vanilla ubuntu

Moved away from a texlive docker container to a vanilla ubuntu runner
so that we can better track the actual dependencies a user needs to
build the PDF rather than relying on the gigabytes of random stuff
that the texlive docker container pulls in.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben committed:
ci: doc: github: drop unused env variable

We get West from the action-zephyr-setup action

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben commented:
Ended up doing lots of houskeeping on the PDF job so as to get rid of the texlive docker container we were using so that we actually install the required packages and therefore cross-check that our documentation has up-to-date instructions w.r.t what people might need to install to build the PDF locally

kartben committed:
doc: ci: github: Require hashes during pip install

require hashes in doc ci workflow

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben commented:
Just added one more extra flag that makes extra sure hashes are mandatory (and makes scorecard happy as well)
